c/src/ml-api-service-remote.c

```diff
@@ -575,70 +578,81 @@ ml_service_remote_release_internal (ml_service_s * mls)
+        path);
+  }
+  if (g_access (path, W_OK) != 0) {
+    _ml_error_report_return (ML_ERROR_PERMISSION_DENIED,
+        "Write permission denied, path: %s", path);
+  }
+
+  g_free (mlrs->path);
+  mlrs->path = g_strdup (path);
+
+  return ML_ERROR_NONE;
+}
+
 /**
  * @brief Creates ml-service handle with given ml-option handle.
  */
 int
-ml_service_remote_create (ml_option_h option, ml_service_event_cb cb,
-    void *user_data, ml_service_h * handle)
+ml_service_remote_create (ml_service_h handle, ml_option_h option)
 {
   ml_service_s *mls;
   _ml_remote_service_s *remote_s = NULL;
   edge_info_s *edge_info = NULL;
   int ret = ML_ERROR_NONE;
-  gchar *_path = NULL;
+  gchar *_path;
 
   check_feature_state (ML_FEATURE_SERVICE);
-  check_feature_state (ML_FEATURE_INFERENCE);
-
-  if (!option) {
-    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
-        "The parameter, 'option' is NULL. It should be a valid ml_option_h, which should be created by ml_option_create().");
-  }
 
   if (!handle) {
     _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
         "The parameter, 'handle' (ml_service_h), is NULL. It should be a valid ml_service_h.");
   }
 
-  if (ML_ERROR_NONE == ml_option_get (option, "path", (void **) (&_path))) {
-    if (!g_file_test (_path, G_FILE_TEST_IS_DIR)) {
-      _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
-          "The given param, dir path = \"%s\" is invalid or the dir is not found or accessible.",
-          _path);
-    }
-    if (g_access (_path, W_OK) != 0) {
-      _ml_error_report_return (ML_ERROR_PERMISSION_DENIED,
-          "Write permission denied, path: %s", _path);
-    }
-  }
-
-  mls = _ml_service_create_internal (ML_SERVICE_TYPE_REMOTE);
-  if (mls == NULL) {
-    _ml_error_report_return (ML_ERROR_OUT_OF_MEMORY,
-        "Failed to allocate memory for the service handle. Out of memory?");
+  if (!option) {
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'option' is NULL. It should be a valid ml_option_h, which should be created by ml_option_create().");
   }
+  mls = (ml_service_s *) handle;
 
   mls->priv = remote_s = g_try_new0 (_ml_remote_service_s, 1);
   if (remote_s == NULL) {
-    _ml_service_destroy_internal (mls);
     _ml_error_report_return (ML_ERROR_OUT_OF_MEMORY,
         "Failed to allocate memory for the service handle's private data. Out of memory?");
   }
 
-  _mlrs_get_edge_info (option, &edge_info);
+  remote_s->table =
+      g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+  if (!remote_s->table) {
+    _ml_error_report
+        ("Failed to allocate memory for the table of ml-service remote. Out of memory?");
+  }
 
-  remote_s->node_type = edge_info->node_type;
-  remote_s->event_cb = cb;
-  remote_s->user_data = user_data;
-  remote_s->path = g_strdup (_path);
+  if (ML_ERROR_NONE == ml_option_get (option, "path", (void **) (&_path))) {
+    ret = ml_service_remote_set_path (mls, _path);
```

Comment:
Readability and stability seem to have improved by making `ml_service_remote_set_path` function 😄